### PR TITLE
feat(graindoc): Allow doc comments on variants and record fields

### LINKED ISSUE #992: GrainDoc: Allow documenting Enum variants
possibly something like such.
```
/*
 * Directions
 *
 * @variant NORTH: The north direction
 * @variant SOUTH: The south direction
 * @variant EAST: The east direction
 * @variant WEST: The west direction
 * /
enum DIRECTION {
  NORTH,
  SOUTH,
  EAST,
  WEST
}
```
and generated markdown
```
| Name    |        Description        |
| ------- | ----------------------    |
| North   | The north direction       |
| South   | The south direction       |
| East    | The east direction        |
| West    | The west direction        |
```

[Solves](https://github.com/grain-lang/grain/pull/945#discussion_r729356805_)

### PR DESCRIPTION
**Pull request by alex-snezhko**
Closes #992 

```
provide record R {
  /**
   * A very special number
   */
  x: Number
}

provide enum E {
  /**
   * A thing
   */
  Thing
}
```

### CODE REVIEW

**Comment by phated** [comment]
Excited to look at this! I'm pretty sure you need to add tests now that @spotandjake added the test harness

File: compiler/test/graindoc/types.expected.md
Comment (spotandjake) [inline compiler/test/graindoc/types.expected.md:43]: I question if this should contain the type info in here as well like 
```suggestion
|`Variant2`|(String, Number)|Another variant|
```
Similar to the docs for functions

File: compiler/test/graindoc/types.input.gr
Comment (spotandjake) [inline compiler/test/graindoc/types.input.gr:29]: Can we get a variant testing `Variant3{ t: Number }` it would be nice if we tested nested docs here?

After looking a little closer as your implementation I dot think you handle allowing docs on nested records we might want to handle that somehow. Idk how just yet maybe we append them to the table like `Variant.t` and the description, though that could be confusing.

**Review by spotandjake — COMMENTED** [review]
This looks really good, I love how much it cleaner it makes the docs. I just had two little notes to keep the tables in line with the function tables but besides that it looks good.

Also after this is merged dont forgot to update the website docs on this feature.

File: compiler/test/graindoc/types.expected.md
Comment (alex-snezhko) [reply compiler/test/graindoc/types.expected.md:43]: Sounds like a good idea to have a separate type column for records, although for variants I'm not sure that this is the best way to display this info, perhaps it would be better to just display the whole variant description in the first column like `Variant2(String, Number)`?

Comment (spotandjake) [reply compiler/test/graindoc/types.expected.md:43]: Maybe, I think that it is worth thinking about consistency if we have functions with a separate type column and use a separate type column for records we should probably use a separate column for variants. I think that things might get a little messy with records attached to variants though so this might need some discussion or be something we come back to in another pr later on.

File: compiler/test/graindoc/types.input.gr
Comment (alex-snezhko) [reply compiler/test/graindoc/types.input.gr:29]: Yeah I thought about this case as well and am unsure of the best way to display it; including a nested table in a new column is one option, and although I initially discounted this option it may be better than separating it out into another top-level table as that would cause an awkward reading flow imo :thinking:

**Review by peblair — COMMENTED** [review]
Looks good, but holding off on an approval until @spotandjake's comments are addressed

File: compiler/test/graindoc/types.input.gr
Comment (alex-snezhko) [reply compiler/test/graindoc/types.input.gr:29]: I tried playing around with different formatting styles and I think having inline record data in a nested table seems like an ok option; since there would only ever be one layer of nesting I think it wouldn't be unreasonable. For example the following code could generate this table:
```
provide enum E {
  /**
   * Singleton constr
   */
  S,
  /**
   * Tuple constr
   */
  T(Number),
  /**
   * Record constr
   */
  R {
    /**
     * A number
     */
    x: Number,
    /**
     * A function
     */
    f: Number -> Number
  }
}
```
image

The image was generated with https://markdownlivepreview.com/, but will probably look similar on the website. Thoughts on this display format?

Comment (spotandjake) [reply compiler/test/graindoc/types.input.gr:29]: It looks better than I was expecting but I wonder if anyone else has any better ideas. Before we implement.

Comment (alex-snezhko) [reply compiler/test/graindoc/types.input.gr:29]: I ended up going ahead and implementing it, if there are suggestions for better output styling I'll update

File: compiler/src/diagnostics/comments.re
Comment (phated) [inline compiler/src/diagnostics/comments.re:239]: I don't like passing this through since it is leaking a super hacky workaround

File: compiler/src/utils/markdown.re
Comment (phated) [inline compiler/src/utils/markdown.re:18]: I don't like boolean flags like this, we should think through a real API that is needed if we are putting tables inside

File: stdlib/sys/process.gr
Comment (phated) [inline stdlib/sys/process.gr:125]: Can we get descriptions on these too?

File: compiler/test/graindoc/types.expected.md
Comment (phated) [inline compiler/test/graindoc/types.expected.md:48]: Sorry for the delay here. I don't really like this layout. How about we do this?

````suggestion
Variants:

```grain
Variant1
```
Enum variant

```grain
Variant2(String, Number)
```
Another variant

```grain
Variant3{
  a: Number,
  f: Number => Number,
}
```
Record Variant

Fields:

|name|type|description|
|----|----|-----------|
|`a`|`Number`|Record field a|
|`f`|`Number => Number`|Function|
````

I was also thinking about making these a bulleted list, but I believe that'd be hard while also making them grain-syntax blocks.

File: stdlib/sys/process.md
Comment (phated) [inline stdlib/sys/process.md:84]: It seems that the enum variants without a description aren't included

File: compiler/test/graindoc/types.expected.md
Comment (alex-snezhko) [reply compiler/test/graindoc/types.expected.md:48]: Assuming that record fields are being kept in a table I'm not a huge fan of having enum variants not be in a table. What about something like this?
```
|variant|description|
|------------|-----------|
|`Variant1`|Enum variant|
|`Variant2(String, Number)`|Another variant|
|<code>Variant3{<br>&nbsp;&nbsp;a: Number,<br>&nbsp;&nbsp;f: Number => Number<br>}</code>|Record variant<br><ul><li>`a`: Record field a</li><li>`f`: Function</li></ul>|
```

image

Comment (phated) [reply compiler/test/graindoc/types.expected.md:48]: I don't like them in a table. It's too dense

File: stdlib/sys/process.md
Comment (alex-snezhko) [reply stdlib/sys/process.md:84]: I did this intentionally. Since the full enum definition is already printed I didn't see much of a point to having an entry with just the variant name and no context

Comment (phated) [reply stdlib/sys/process.md:84]: That's fine when we remove the table for variants. If it were a table like you did originally, it should be the full enumeration.

Comment (alex-snezhko) [reply stdlib/sys/process.md:84]: This would also apply to record fields though. Unless we want to ditch the table for those as well?

Comment (phated) [reply stdlib/sys/process.md:84]: Looking at my layout suggestion, record fields remain in a table, which contains all fields.

File: stdlib/sys/process.gr
Comment (alex-snezhko) [reply stdlib/sys/process.gr:125]: On second thought should these just be removed altogether? Looks like signal handling isn't a thing anymore https://github.com/WebAssembly/WASI/issues/166

File: compiler/test/graindoc/types.expected.md
Comment (alex-snezhko) [reply compiler/test/graindoc/types.expected.md]: Applied your suggested format

File: stdlib/sys/process.gr
Comment (alex-snezhko) [reply stdlib/sys/process.gr:125]: I think maybe we can clean these up when we rewrite everything to use wasi preview 2?

File: compiler/graindoc/docblock.re
Comment (phated) [inline compiler/graindoc/docblock.re:246]: Can you flip this logic so we don't need the `!`?

File: compiler/graindoc/docblock.re
Comment (phated) [inline compiler/graindoc/docblock.re:279]: Can you flip the if-else to avoid the `!`?

File: compiler/graindoc/docblock.re
Comment (phated) [inline compiler/graindoc/docblock.re:536]: Does this need the logic for non-duplicate `@since` that @spotandjake added in the other PR?

File: compiler/src/typed/printtyp.re
Comment (phated) [inline compiler/src/typed/printtyp.re:995]: Perhaps open an issue to clean this up when we drop oprint for the formatter code?

**Review by phated — CHANGES_REQUESTED** [review]
I like the output a lot! I had just a couple comments on the implementation

File: compiler/src/typed/printtyp.re
Comment (alex-snezhko) [reply compiler/src/typed/printtyp.re:995]: https://github.com/grain-lang/grain/issues/1992

File: compiler/graindoc/docblock.re
Comment (alex-snezhko) [reply compiler/graindoc/docblock.re:536]: This would raise the exception if there are any attributes on the docblock at all, so deduplicating isn't necessary here

Comment (phated) [reply compiler/graindoc/docblock.re:532]: Hm, I'm pretty sure the JSON PR uses the since attribute on each variant so we'll want to support that

Comment (alex-snezhko) [reply compiler/graindoc/docblock.re:532]: Hmm ok, should `@deprecated` and `@history` be supported as well?

Comment (phated) [reply compiler/graindoc/docblock.re:532]: Yeah, I think those would be supported

Comment (alex-snezhko) [reply compiler/graindoc/docblock.re:532]: What about record fields? And if so then how would they be displayed in the table?

Comment (phated) [reply compiler/graindoc/docblock.re:532]: It's probably fine to only support them on variants. When thinking about exhaustive matching, I think about variants more than record fields and you might add variants over time.

Comment (alex-snezhko) [reply compiler/graindoc/docblock.re:532]: Hmm, I think it'd be confusing if they were supported on variants but not record fields; for instance if you had a field that changed its type or name or something. What if it's just inserted below the field's description in the table cell? It wouldn't take up much space by default since it'd be collapsed in the `<details>` tag

Comment (phated) [reply compiler/graindoc/docblock.re:532]: I don't want it in the table. We can open a follow up issue to evaluate for records

Comment (phated) [reply compiler/graindoc/docblock.re:532]: Merging without attribute support and @spotandjake can remove them from the JSON PR

Comment (phated) [reply compiler/graindoc/docblock.re:532]: #1998 and #1999

**Review by phated — APPROVED** [review]

**Pull request merged**
